### ISSUE TITLE
dbus daemon will die

### ISSUE THREAD

**Issue by sunjoong** [Microsoft/WSL#2339]
* Your Windows build number:  Microsoft Windows [Version 10.0.15063]

* What you're doing and what's happening:

I'm using on Gentoo. I wrote a vbscript to launch lxde;
```
Option Explicit

Dim   Instance
Const VcXsrv = "C:\Program Files\VcXsrv\vcxsrv.exe"
Const PulseAudio = "C:\Users\sunjoong\AppData\Local\PulseAudio\bin\pulseaudio.exe"

CreateObject("Shell.Application").ShellExecute _
  VcXsrv, ":0 -nodecoration -clipboard -wgl", "", "open", 0
CreateObject("Shell.Application").ShellExecute _
  PulseAudio, "--exit-idle-time=-1", "", "open", 0
CreateObject("WScript.Shell").Run _
  "bash.exe ~ -c 'export DISPLAY=:0; " & _
                 "export PULSE_SERVER=tcp:127.0.0.1; " & _
                 "source /etc/profile; " & _
                 "sudo mkdir -p /var/run/dbus; " & _
                 "sudo /usr/bin/dbus-daemon --system; " & _
                 "sudo /usr/bin/dbus-send --print-reply --system --type=method_call --dest=org.freedesktop.DBus / org.freedesktop.DBus.ReloadConfig; " & _
                 "sudo mkdir -p -m 0755 /run/ConsoleKit; " & _
                 "sudo /usr/sbin/console-kit-daemon; " & _
                 "startlxde'", _
  0, 1
For each Instance in GetObject ("winmgmts:").InstancesOf("Win32_Process")
  If StrComp(Instance.ExecutablePath, PulseAudio, 1) = 0 then
    Instance.Terminate()
    exit For
  end If
Next
For each Instance in GetObject ("winmgmts:").InstancesOf("Win32_Process")
  If StrComp(Instance.ExecutablePath, VcXsrv, 1) = 0 then
    Instance.Terminate()
    exit For
  end If
Next
Set Instance = Nothing

WScript.Quit
```
After launching lxde, process list was;
```
sunjoong@SUNJOONG-DESKTOP ~ $ ps -ax
  PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /init
    2 tty1     Ss     0:00 /init
    3 tty1     S      0:00 /bin/bash -c export DISPLAY=:0; export PULSE_SERVER=tcp:127.0.0.1; source /etc/profile; sudo mkdir -p /var/run/dbus; sudo /usr/bin/dbus-daemon --system; 
    6 ?        Ss     0:00 /usr/bin/dbus-daemon --system
   10 ?        Ssl    0:00 /usr/sbin/console-kit-daemon
   11 tty1     Sl     0:00 /usr/bin/lxsession -s LXDE -e LXDE
   18 ?        Sl     0:00 /usr/lib/polkit-1/polkitd --no-debug
   26 tty1     S      0:00 dbus-launch --sh-syntax --exit-with-session
   30 ?        Ss     0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
   35 tty1     S      0:00 openbox --config-file /home/sunjoong/.config/openbox/lxde-rc.xml
   36 tty1     Sl     0:00 lxpolkit
   38 tty1     Sl     0:00 lxpanel --profile LXDE
   39 tty1     Sl     0:00 pcmanfm --desktop --profile LXDE
   42 tty1     S      0:00 lxclipboard
   44 tty1     S      0:00 /usr/libexec/polkit-gnome-authentication-agent-1
   49 ?        S      0:00 /usr/bin/dbus-daemon --system
   50 ?        Sl     0:00 /usr/sbin/console-kit-daemon --no-daemon
   57 ?        S      0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
   58 ?        Sl     0:00 /usr/libexec/at-spi-bus-launcher
   65 ?        Ssl    0:00 /usr/libexec/menu-cache/menu-cached /tmp/.menu-cached-:0-sunjoong
   74 tty1     Sl     0:00 lxterminal
   75 tty1     S      0:00 gnome-pty-helper
   76 pts/0    Ss     0:00 /bin/bash
   80 pts/0    R      0:00 ps -ax
sunjoong@SUNJOONG-DESKTOP ~ $ 
```
Then, "`Timeout was reached`" popup was displayed. After that, process list was;
```
sunjoong@SUNJOONG-DESKTOP ~ $ ps -ax
  PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /init
    2 tty1     Ss     0:00 /init
    3 tty1     S      0:00 /bin/bash -c export DISPLAY=:0; export PULSE_SERVER=tcp:127.0.0.1; source /etc/profile; sudo mkdir -p /var/run/dbus; sudo /usr/bin/dbus-daemon --system; 
   10 ?        Ssl    0:00 /usr/sbin/console-kit-daemon
   11 tty1     Sl     0:00 /usr/bin/lxsession -s LXDE -e LXDE
   26 tty1     S      0:00 dbus-launch --sh-syntax --exit-with-session
   30 ?        Ss     0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
   35 tty1     S      0:00 openbox --config-file /home/sunjoong/.config/openbox/lxde-rc.xml
   36 tty1     Sl     0:00 lxpolkit
   38 tty1     Sl     0:00 lxpanel --profile LXDE
   39 tty1     Sl     0:00 pcmanfm --desktop --profile LXDE
   42 tty1     S      0:00 lxclipboard
   50 ?        Sl     0:00 /usr/sbin/console-kit-daemon --no-daemon
   57 ?        S      0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
   58 ?        Sl     0:00 /usr/libexec/at-spi-bus-launcher
   65 ?        Ssl    0:00 /usr/libexec/menu-cache/menu-cached /tmp/.menu-cached-:0-sunjoong
   74 tty1     Sl     0:00 lxterminal
   75 tty1     S      0:00 gnome-pty-helper
   76 pts/0    Ss     0:00 /bin/bash
   81 pts/0    R      0:00 ps -ax
sunjoong@SUNJOONG-DESKTOP ~ $ 
```
In short, these processes were dead and gone;
```
    6 ?        Ss     0:00 /usr/bin/dbus-daemon --system
   18 ?        Sl     0:00 /usr/lib/polkit-1/polkitd --no-debug
   44 tty1     S      0:00 /usr/libexec/polkit-gnome-authentication-agent-1
   49 ?        S      0:00 /usr/bin/dbus-daemon --system
```
lxsession have dependency on polkit on gentoo but polkit might make a problem, I suspect.

* Console case

I opened bash console and ran dbus-daemon and console-kit manually;
```
root@SUNJOONG-DESKTOP:~# ps -ax
  PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /init
    2 tty1     Ss     0:00 /bin/bash
   11 ?        Ss     0:00 /usr/bin/dbus-daemon --system
   15 ?        Ssl    0:00 /usr/sbin/console-kit-daemon
   21 ?        Sl     0:00 /usr/lib/polkit-1/polkitd --no-debug
   27 tty1     R      0:00 ps -ax
root@SUNJOONG-DESKTOP:~#
```
In this, console-kit spawned polkit and dbus-daemon and polkitd were alive.

* Ubuntu case

On gentoo, there was no error popup to begin with and then "`Timeout was reached`" popup after time went by. On ubuntu, there was "`No session for pid 18`" popup from first.
```
root@SUNJOONG-DESKTOP:~# ps -ax
  PID TTY      STAT   TIME COMMAND
    1 ?        Ss     0:00 /init
    2 tty1     Ss     0:00 /init
    3 tty1     S      0:00 /bin/bash -c export DISPLAY=:0; export PULSE_SERVER=tcp:127.0.0.1; source /etc/profile; export LC_ALL=C;sudo mkdir -p /var/run/dbu
   10 ?        Ss     0:00 /usr/bin/dbus-daemon --system
   17 ?        Ssl    0:00 /usr/sbin/console-kit-daemon
   18 tty1     Sl     0:00 /usr/bin/lxsession -s LXDE -e LXDE
   21 ?        Ss     0:00 /usr/bin/ssh-agent -s
   28 tty1     S      0:00 dbus-launch --sh-syntax --exit-with-session
   29 ?        Ss     0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
   35 ?        Sl     0:00 /usr/lib/gvfs/gvfsd
   45 tty1     S      0:00 openbox --config-file /root/.config/openbox/lxde-rc.xml
   50 tty1     Sl     0:00 lxpanel --profile LXDE
   51 tty1     Sl     0:00 pcmanfm --desktop --profile LXDE
   52 tty1     S      0:00 xscreensaver -no-splash
   59 tty1     Sl     0:00 nm-applet
   61 ?        Ss     0:00 /usr/bin/ssh-agent -s
   63 tty1     Sl     0:00 /usr/lib/at-spi2-core/at-spi-bus-launcher --launch-immediately
   65 tty1     Sl     0:00 clipit
   80 tty1     S      0:00 /usr/bin/dbus-daemon --config-file=/etc/at-spi2/accessibility.conf --nofork --print-address 3
   90 ?        Ssl    0:00 /usr/lib/menu-cache/menu-cached /tmp/.menu-cached-:0-root
   96 ?        Sl     0:00 /usr/lib/gvfs/gvfs-udisks2-volume-monitor
  102 tty1     Sl     0:00 /usr/lib/at-spi2-core/at-spi2-registryd --use-gnome-session
  111 ?        Sl     0:00 /usr/lib/udisks2/udisksd --no-debug
  120 ?        Sl     0:00 /usr/lib/policykit-1/polkitd --no-debug
  133 ?        Sl     0:00 /usr/lib/gvfs/gvfs-mtp-volume-monitor
  140 ?        Sl     0:00 /usr/lib/gvfs/gvfs-gphoto2-volume-monitor
  147 ?        Sl     0:00 /usr/lib/gvfs/gvfs-goa-volume-monitor
  153 ?        Sl     0:00 /usr/lib/gvfs/gvfs-afc-volume-monitor
  162 ?        Sl     0:00 /usr/lib/gvfs/gvfsd-trash --spawner :1.1 /org/gtk/gvfs/exec_spaw/0
  169 tty1     Sl     0:00 lxterminal
  170 tty1     S      0:00 gnome-pty-helper
  171 pts/0    Ss     0:00 /bin/bash
  187 pts/0    R      0:00 ps -ax
root@SUNJOONG-DESKTOP:~# 
```

* PIDs

On gentoo, dbus-daemon had PID 6 and 49;
```
    6 ?        Ss     0:00 /usr/bin/dbus-daemon --system
   49 ?        S      0:00 /usr/bin/dbus-daemon --system
```
But, it had been 5;
```
sunjoong@SUNJOONG-DESKTOP ~ $ cat /var/run/dbus.pid 
5
sunjoong@SUNJOONG-DESKTOP ~ $ 
```
And, there were traces on gentoo;
```
   30 ?        Ss     0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
   57 ?        S      0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
```
and on ubuntu too;
```
   29 ?        Ss     0:00 /usr/bin/dbus-daemon --fork --print-pid 5 --print-address 7 --session
```

* Error message popup

On gentoo, there was  "`Timeout was reached`" popup; I geuss lxsession spawned polkit and it will reached timeout, but polkit itself for new session seems to work from on console case result.

On ubuntu, there was "`No session for pid 18`" popup; I guess lxsession will not send a message.

* Any idea?

I want to get rid of  "`Timeout was reached`" popup and make dbus work well. Any idea?

__EDIT__: On console, PID of dbus.pid did not die, I found. And... I collected logs but I'm not sure wheiter it was done with right way; [strace.zip](https://github.com/Microsoft/BashOnWindows/files/1151237/strace.zip). Log of lxpolkit is on startlxde.45 file.

**Comment by shevegen** [comment]
In my experience so far, which is quite limited (~4 weeks at best so far with win10), dbus in general does not work extremely well. Since it is quite important on modern *nix systems, hopefully things can be improved. If all else fails, perhaps the dbus maintainers could ... write their code a little bit better so that it works better, too. I can run editors such as geany (via Xming) now but dbus still fails for me ...  :P

**Comment by sunjoong** [comment]
@shevegen - Thanks for comment  :). I was avoiding dbus as far as possible, but began to think I might need dbus for more fancy softwares few days ago. I hope to be able to find a workaround or settings for dbus.

**Comment by fox8091** [comment]
@sunjoong Perhaps change dbus to work over localhost TCP? A shell command to do it is `sudo sed -i 's$<listen>.*</listen>$<listen>tcp:host=localhost,port=0</listen>$' /etc/dbus-1/session.conf`.

**Comment by sunjoong** [comment]
@fox8091 - Thans to remind me. I had put session-local.conf file into /etc/dbus-1/ directory,
```
<busconfig>
<listen>tcp:host=localhost,port=0,family=ipv4</listen>
<auth>ANONYMOUS</auth>
<allow_anonymous/>
</busconfig>
```
because there is "`<include ignore_missing="yes">/etc/dbus-1/session-local.conf</include>`" line in /usr/share/dbus-1/session.conf file. So... Yes, I had configured "dbus to work over localhost TCP."

**Comment by sunilmut** [comment]
@sunjoong - Thanks for your post. Is there a more targeted repro for this? That is easier for us to go through/troubleshoot/triage.

We are aware that there are still some limitations (as also noted by @shevegen) in `dbus` (for ex: the dbus test does not fully pass) and we would like to get dbus to a better place on WSL as well.

**Comment by sunjoong** [comment]
@sunilmut - Thanks for reply. Hmm... more targeted repro... I'm afraid I don't get it. Well... instead, I'll say the background of that.

One day, I had thought it could be cool to use a web browser on WSL when I had needed a new web brower for some site that does not support Microsoft Edge and Internet Explorer 11 on Windows 10. Of course, installing Google Chrome on Windows could be a solution but I had thought why not on WSL.

I had searched and known Chrome (and maybe Chromium also) does not work on WSL yet, so aimed Firefox. It had worked on that site so I had gone further. It could play youtube without sound, so I had recompile it with pulseaudio support but it had frozen. I had thought dbus could make it, and found dbus daemon of pid 5 will die. I'm not sure that might be related to my aim but looks weird.

I heard WSL does not support sound yet but we have a workaround of running pulseaudio server on Windows. I hope dbus on WSL work well for my aim but... Hmm... Is there any workaround using dbus on Windows and communication with programs on WSL? I found dbus on Windows on https://code.google.com/archive/p/dbus-windows-installer/ and could launch dbus-daemon.
